src/Shared/Buffers.MemoryPool/UnmanagedBufferWriter.cs

```diff
@@ -0,0 +1,95 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+
+/// <summary>
+/// An <see cref="IBufferWriter{T}"/> implementation for writing to
+/// unmanaged blocks of memory.
+/// </summary>
+internal sealed unsafe class UnmanagedBufferWriter : IBufferWriter<byte>, IDisposable
```

Comment:
Does this really need to go through `IBufferWriter<byte>` abstraction? I think this can be (ref) struct and it does not need to implement `IBufferWriter<byte>` interface. It would save allocation and some interface virtual calls.

Reply:
Agree we should be able to get away with something like https://github.com/dotnet/aspnetcore/blob/75ada08ac7763cf174c05fbe0755e184c0ad9737/src/Shared/ServerInfrastructure/BufferWriter.cs. You can avoid the IBufferWriter calls for encoding as well and just pass the `Span<byte>`